ext/bcmath/libbcmath/src/recmul.c

```diff
@@ -277,34 +351,424 @@ static void bc_standard_mul(bc_num n1, size_t n1len, bc_num n2, size_t n2len, bc
+ * to about N^1.585.
+ */
+static void bc_rec_mul(bc_num n1, size_t n1len, bc_num n2, size_t n2len, bc_num *prod)
+{
+	size_t i;
+	const char *n1end = n1->n_value + n1len - 1;
+	const char *n2end = n2->n_value + n2len - 1;
+	size_t prodlen = n1len + n2len;
+
+	size_t n1_arr_size = (n1len + BC_MUL_UINT_DIGITS - 1) / BC_MUL_UINT_DIGITS;
+	size_t n2_arr_size = (n2len + BC_MUL_UINT_DIGITS - 1) / BC_MUL_UINT_DIGITS;
+	/*
+	 * For computational efficiency, the size of prod_uint is slightly larger.
+	 * Therefore, calculate the actual required size to avoid overflow when converting to BCD.
+	 */
+	size_t prod_arr_real_size = n1_arr_size + n2_arr_size - 1;
+
+	/*
+	 * Adjust size to a multiple of 2 for computational efficiency.
+	 */
+	n1_arr_size = (n1_arr_size + 1) & ~1;
+	n2_arr_size = (n2_arr_size + 1) & ~1;
+	size_t prod_arr_size = n1_arr_size + n2_arr_size;
+
+	/*
+	 * Find the next largest power of 2 for n1_arr_size and n2_arr_size.
+	 * Use the larger one as calc_size. This is the computational length for the
+	 * divide and conquer calculation. e.g. For 7 * 3, the powers of 2 are 8 and 4,
+	 * and the larger one, 8, is the calc_size.
+	 *
+	 * Each value found here is used to calculate the buffer size to hold the
+	 * subtraction result of the divide-and-conquer mid calculation (so the variables
+	 * are reused).
+	 *
+	 * How to calculate the buffer size is explained in the memory allocation comments.
+	 */
+	size_t calc_size;
+	size_t n1_buf_size = bc_mul_next_pow_2(n1_arr_size);
+	size_t n2_buf_size = bc_mul_next_pow_2(n2_arr_size);
+	if (n1_buf_size > n2_buf_size) {
+		/*
+		 * If arr_size is small, size adjustment will be performed. See the memory
+		 * allocation comments for more details.
+		 */
+		n2_buf_size += n2_arr_size * (n1_buf_size / n2_buf_size - 1);
+		calc_size = n1_buf_size;
+	} else if (n1_buf_size < n2_buf_size) {
+		/*
+		 * If arr_size is small, size adjustment will be performed. See the memory
+		 * allocation comments for more details.
+		 */
+		n1_buf_size += n1_arr_size * (n2_buf_size / n1_buf_size - 1);
+		calc_size = n2_buf_size;
+	} else {
+		calc_size = n1_buf_size;
+	}
+
+	/*
+	 * How to calculate memory size
+	 *
+	 * n1_uint, n2_uint:
+	 * Use arr_size. The sum of these is prod_arr_size.
+	 *
+	 * n1_buf, n2_buf:
+	 * These are the necessary buffers for the divide and conquer "mid".
+	 * Assuming the computation length is M, the first split uses a buffer of M/2, and the second
+	 * split uses a buffer of M/4.... The required buffer size increases by half until the size
+	 * reaches 2. In other words, it is the sum of a geometric progression of 2 with a geometric
+	 * ratio of 2 from 2 to N-1, where the calculation length is N.
+	 * This sum can be calculated using the following formula, where the first term is a and the
+	 * geometric ratio is r: a(r^(N-1) - 1)(r - 1)
+	 * Here, a and r are both 2, so this formula becomes: 2(2^(N-1) - 1)(2 - 1) = 2^N - 2
```

Comment:
```suggestion
	 * Here, a and r are both 2, so this formula becomes: 2(2^(N-1) - 1)/(2 - 1) = 2^N - 2
```

Reply:
Why are a and r both 2? I would've expected `a = M` and `r = 1/2`

Reply:
It was intended that rearranging the numbers in reverse would result in a sequence like the comments, but the format you've written may be easier to understand.